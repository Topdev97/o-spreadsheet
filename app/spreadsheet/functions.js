export const functions = {
  SUM: {
    description: "Returns the sum of all values in a range",
    compute: function(range) {
      return range.reduce((a, b) => a + b, 0);
    }
  }
};
