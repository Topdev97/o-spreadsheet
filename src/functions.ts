export const functions = {
  SUM: {
    description: "Returns the sum of all values in a range",
    compute: function(...args) {
      return args.flat().reduce((a, b) => a + b, 0);
    }
  },
  RAND: {
    description: "Returns a random number between 0 and 1",
    compute: function() {
      return Math.random();
    }
  }
};
