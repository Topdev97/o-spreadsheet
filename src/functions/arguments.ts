export type ArgType = "BOOLEAN" | "ANY" | "RANGE" | "NUMBER" | "STRING";

export interface Arg {
  repeating?: boolean;
  optional?: boolean;
  description: string;
  name: string;
  type: ArgType[];
  default?: any;
  valueProvider?: Function;
  isValueProviderRestrictive?: boolean;
}

//------------------------------------------------------------------------------
// Argument validation
//------------------------------------------------------------------------------

export function validateArguments(args: Arg[]) {
  let previousArgRepeating: boolean | undefined = false;
  let previousArgOptional: boolean | undefined = false;
  for (let current of args) {
    if (previousArgRepeating) {
      throw new Error(
        "Function ${name} has at least 2 arguments that are repeating. The maximum repeating arguments is 1."
      );
    }

    if (previousArgOptional && !current.optional) {
      throw new Error(
        "Function ${name} has at mandatory arguments declared after optional ones. All optional arguments must be after all mandatory arguments."
      );
    }
    previousArgRepeating = current.repeating;
    previousArgOptional = current.optional;
  }
}

//------------------------------------------------------------------------------
// Wrapping functions for arguments sanitization
//------------------------------------------------------------------------------
export function protectFunction(fn: Function, argList: Arg[]): Function {
  if (argList.length === 0) {
    return fn;
  }
  return function(this: any, ...args: any[]) {
    args = sanitizeArgs(args, argList);
    return fn.call(this, ...args);
  };
}

/**
 * If you read this, and are horrified by the code, worry not, dear friend.
 * This code is here only for a while, to solidify the test suite and prepare
 * the future. It will be replaced by a shiny argument sanitizer compiler soon.
 *
 * Note: this function modifies args in place!
 */
export function sanitizeArgs(args: any[], argList: Arg[]): any[] {
  for (let i = 0; i < argList.length; i++) {
    const descr = argList[i];
    if (!(i in args) && !descr.optional) {
      throw new Error("Wrong number of arguments. Expected 1, but got 0 argument instead.");
    }
    const arg = args[i];
    if (arg === undefined && descr.optional) {
      args = args.slice(0, i);
      break;
    }
    if (descr.repeating) {
      for (let j = i; j < args.length; j++) {
        sanitizeArg(args, j, args[j], descr);
      }
      return args;
    }
    sanitizeArg(args, i, arg, descr);
  }
  if (args.length > argList.length) {
    throw new Error(
      `Wrong number of arguments. Expected ${argList.length}, but got ${args.length} arguments instead.`
    );
  }
  return args;
}

function sanitizeArg(args: any[], i: number, arg: any, descr: Arg) {
  if (arg === undefined) {
    if (descr.type.includes("NUMBER")) {
      args[i] = 0;
    }
    if (descr.type.includes("BOOLEAN")) {
      args[i] = false;
    }
  } else if (typeof arg === "boolean" && !descr.type.includes("BOOLEAN")) {
    if (descr.type.includes("NUMBER")) {
      args[i] = arg ? 1 : 0;
    }
  } else if (typeof arg === "string") {
    if (descr.type.includes("NUMBER")) {
      if (arg) {
        const n = Number(arg);
        if (isNaN(n)) {
          throw new Error(
            `Argument "${descr.name}" should be a number, but "${arg}" is a text, and cannot be coerced to a number.`
          );
        }
      } else {
        args[i] = 0;
      }
      args[i] = arg ? parseFloat(arg) : 0;
    } else if (descr.type.includes("BOOLEAN")) {
      if (arg === "") {
        args[i] = false;
      } else if (arg.toUpperCase() === "TRUE") {
        args[i] = true;
      } else if (arg.toUpperCase() === "FALSE") {
        args[i] = false;
      } else {
        throw new Error(
          `Argument "${descr.name}" should be a boolean, but "${arg}" is a text, and cannot be coerced to a boolean.`
        );
      }
    }
  } else if (typeof arg === "number") {
    if (descr.type.includes("BOOLEAN")) {
      args[i] = arg ? true : false;
    }
  }
}
//------------------------------------------------------------------------------
// Arg description DSL
//------------------------------------------------------------------------------

const ARG_REGEXP = /(.*)\((.*)\)(.*)/;
const ARG_TYPES: ArgType[] = ["BOOLEAN", "NUMBER", "STRING", "ANY", "RANGE"];

/**
 * This function is meant to be used as a tag for a template strings.
 *
 * Its job is to convert a textual description of the list of arguments into an
 * actual array of Arg, suitable for consumption.
 */
export function args(strings: TemplateStringsArray): Arg[] {
  let lines = strings[0].split("\n");
  const result: Arg[] = [];
  for (let l of lines) {
    l = l.trim();
    if (l) {
      result.push(makeArg(l));
    }
  }
  return result;
}

function makeArg(str: string): Arg {
  let parts = str.match(ARG_REGEXP)!;
  let name = parts[1].trim();
  let types: ArgType[] = [];
  let isOptional = false;
  let isRepeating = false;
  for (let param of parts[2].split(",")) {
    param = param.trim().toUpperCase();
    let type = ARG_TYPES.find(t => param === t);
    if (type) {
      types.push(type);
    } else if (param === "OPTIONAL") {
      isOptional = true;
    } else if (param === "REPEATING") {
      isRepeating = true;
    }
  }
  let description = parts[3].trim();
  const result: Arg = {
    name,
    description,
    type: types
  };
  if (isOptional) {
    result.optional = true;
  }
  if (isRepeating) {
    result.repeating = true;
  }
  return result;
}
