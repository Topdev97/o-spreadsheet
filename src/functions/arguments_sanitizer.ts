import { Arg } from "./arguments";

const PRIMITIVE_TYPES = ["NUMBER", "STRING", "BOOLEAN"];
const RANGE_TYPES = ["RANGE<NUMBER>", "RANGE<STRING>", "RANGE<BOOLEAN>"];

export function makeSanitizer(args: Arg[]): Function {
  const code: string[] = [];
  const hasRepeatingArg = args.length && args[args.length - 1].repeating;

  const min = args.filter(a => !a.optional).length;
  const max = args.length;
  if (min === max && !hasRepeatingArg) {
    code.push(`if (arguments.length !== ${args.length}) {`);
    code.push(
      `  throw new Error(\`Wrong number of arguments. Expected ${args.length}, but got \$\{arguments.length\} argument(s) instead.\`);`
    );
    code.push(`}`);
  } else {
    const conditions: string[] = [];
    if (min > 0) {
      conditions.push(`arguments.length < ${min}`);
    }
    if (!hasRepeatingArg) {
      conditions.push(`arguments.length > ${args.length}`);
    }
    if (conditions.length) {
      code.push(`if (${conditions.join(" || ")}) {`);
      code.push(
        `  throw new Error(\`Wrong number of arguments. Expected ${args.length}, but got \$\{arguments.length\} argument(s) instead.\`);`
      );
      code.push(`}`);
    }
  }
  const argNames: string[] = [];
  for (let i = 0; i < args.length; i++) {
    let name = `arg${i}`;
    argNames.push(args[i].repeating ? `...${name}` : `${name}`);
    sanitizeArg(code, args[i], name);
  }
  code.push(`return this.fn(${argNames.join(",")});`);
  const sanitizer = new Function(...argNames, code.join("\n"));
  return sanitizer;
}

function sanitizeArg(code: string[], arg: Arg, name: string) {
  if (arg.type.filter(t => PRIMITIVE_TYPES.includes(t)).length > 1) {
    throw new Error("Unsupported type definition. Please go talk to GED about this");
  }
  if (arg.repeating) {
    code.push(`for (let i = 0; i < ${name}.length; i++) {`);
    const basicArg = Object.assign({}, arg, { repeating: false, optional: false });
    code.push(`let ${name}_value = ${name}[i];`);
    sanitizeArg(code, basicArg, name + "_value");
    code.push(`${name}[i] = ${name}_value;`);
    code.push(`}`);
    return;
  }
  const rangeType = arg.type.find(t => RANGE_TYPES.includes(t));
  const type = rangeType ? rangeType.slice(6, -1).toLowerCase() : "";

  if (arg.type.includes("NUMBER")) {
    code.push(`switch (typeof ${name}) {`);
    if (!arg.optional || (arg.optional && arg.default)) {
      code.push(` case "undefined":`);
      code.push(`   ${name} = ${arg.default || 0};`);
      code.push(`   break;`);
    }
    code.push(` case "boolean":`);
    code.push(`   ${name} = ${name} ? 1 : 0;`);
    code.push(`   break;`);
    code.push(` case "string":`);
    code.push(`   if (${name}) {`);
    code.push(`     let n = Number(${name});`);
    code.push(`     if (isNaN(n)) {`);
    code.push(
      `       throw new Error(\`Argument "${arg.name}" should be a number, but "\$\{${name}\}" is a text, and cannot be coerced to a number.\`);`
    );
    code.push(`     } else {`);
    code.push(`       ${name} = n;`);
    code.push(`     }`);
    code.push(`   } else {`);
    code.push(`     ${name} = 0;`);
    code.push(`   }`);
    code.push(`   break;`);

    if (rangeType) {
      code.push(`  case "object":`);
      sanitizeRange(code, name, arg, type);
      code.push(`    break;`);
    }
    code.push(`}`);
    return;
  }

  if (arg.type.includes("BOOLEAN")) {
    code.push(`switch (typeof ${name}) {`);
    code.push(` case "undefined":`);
    code.push(`   ${name} = false;`);
    code.push(`   break;`);
    code.push(` case "number":`);
    code.push(`   ${name} = ${name} ? true : false;`);
    code.push(`   break;`);
    code.push(` case "string":`);
    code.push(`   if (${name}) {`);
    code.push(`     let uppercaseVal = ${name}.toUpperCase();`);
    code.push(`     if (uppercaseVal === "TRUE") {`);
    code.push(`       ${name} = true;`);
    code.push(`     } else if (uppercaseVal === "FALSE") {`);
    code.push(`       ${name} = false;`);
    code.push(`     } else {`);
    code.push(
      `       throw new Error(\`Argument "${arg.name}" should be a boolean, but "\$\{${name}\}" is a text, and cannot be coerced to a boolean.\`);`
    );
    code.push(`     }`);
    code.push(`   } else {`);
    /**
     * @compatibility Note: this is not the way Google Sheets behave:
     *
     * =if("", 1, 2) is evaluated to 2
     * =or("", 1) throws an error
     *
     * It is not clear (to me) why in the first expression it looks like it
     * is accepted, but not in the second.
     */
    code.push(`     ${name} = false;`);
    code.push(`   }`);
    code.push(`   break;`);
    code.push(`}`);
  }

  if (arg.type.includes("STRING")) {
    code.push(`switch (typeof ${name}) {`);
    code.push(` case "undefined":`);
    code.push(`   ${name} = "";`);
    code.push(`   break;`);
    code.push(` case "number":`);
    code.push(`   ${name} = ${name}.toString();`);
    code.push(`   break;`);
    code.push(` case "boolean":`);
    code.push(`   ${name} = ${name} ? "TRUE" : "FALSE";`);
    code.push(`   break;`);
    code.push(`}`);
  }

  if (arg.type.includes("RANGE")) {
    code.push(`if (!(${name} instanceof Array)) {`);
    code.push(`   throw new Error(\`Argument "${arg.name}" has the wrong type\`);`);
    code.push(`}`);
  }

  if (rangeType) {
    sanitizeRange(code, name, arg, type);
  }
}

function sanitizeRange(code: string[], name: string, arg: Arg, type: string) {
  code.push(`if (!(${name} instanceof Array)) {`);
  code.push(`   throw new Error(\`Argument "${arg.name}" has the wrong type\`);`);
  code.push(`}`);
  code.push(`for (let i = 0; i < ${name}.length; i++) {`);
  code.push(`  let col = ${name}[i];`);
  code.push(`  for (let j = 0; j < col.length; j++) {`);
  code.push(`    if (typeof col[j] !== "${type}") {`);
  code.push(`      col[j] = undefined;`);
  code.push(`    }`);
  code.push(`  }`);
  code.push(`}`);
}
//------------------------------------------------------------------------------
// Wrapping functions for arguments sanitization
//------------------------------------------------------------------------------
export function protectFunction(fn: Function, argList: Arg[]): Function {
  if (argList.length === 0) {
    return fn;
  }
  const sanitizer = makeSanitizer(argList);
  return sanitizer.bind({ fn });
}
