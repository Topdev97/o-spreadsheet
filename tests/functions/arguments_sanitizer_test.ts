import { protectFunction, makeSanitizer } from "../../src/functions/arguments_sanitizer";
import { args, Range } from "../../src/functions/arguments";

describe("protectFunction", () => {
  test("function with no args is not wrapped", () => {
    function fn() {}
    expect(protectFunction(fn, [])).toBe(fn);
  });
});

describe("makeSanitizer", () => {
  test("a single number argument", () => {
    const argList = args`n (number) some number`;

    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();
    // const sanitizeArgs = transformSanitizer(sanitizer);
    expect(sanitizer([1])).toEqual([1]);
    expect(sanitizer([false])).toEqual([0]);
    expect(sanitizer([true])).toEqual([1]);
    expect(sanitizer([undefined])).toEqual([0]);
    expect(sanitizer([""])).toEqual([0]);
    expect(sanitizer(["1"])).toEqual([1]);
    expect(sanitizer(["-1"])).toEqual([-1]);
    expect(sanitizer(["1.1"])).toEqual([1.1]);

    expect(() => sanitizer(["ab"])).toThrow(
      'Argument "n" should be a number, but "ab" is a text, and cannot be coerced to a number.'
    );
    expect(() => sanitizer([])).toThrow(
      "Wrong number of arguments. Expected 1, but got 0 argument(s) instead."
    );
    expect(() => sanitizer([1, 2])).toThrow(
      "Wrong number of arguments. Expected 1, but got 2 argument(s) instead."
    );
  });

  test("an optional number argument", () => {
    const argList = args`n (number,optional) some number`;

    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    expect(sanitizer([1])).toEqual([1]);
    expect(sanitizer([false])).toEqual([0]);
    expect(sanitizer([true])).toEqual([1]);
    expect(sanitizer([undefined])).toEqual([undefined]);
    expect(sanitizer([])).toEqual([]);
  });

  test("an optional number argument with a default value", () => {
    const argList = args`n (number,optional,default=42) some number`;

    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    expect(sanitizer([1])).toEqual([1]);
    expect(sanitizer([false])).toEqual([0]);
    expect(sanitizer([true])).toEqual([1]);
    expect(sanitizer([undefined])).toEqual([42]);
    expect(sanitizer([])).toEqual([42]);
  });

  test("repeating, non optional, number argument", () => {
    const argList = args`n (number,repeating) some number`;

    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    expect(() => sanitizer([])).toThrow(
      "Wrong number of arguments. Expected 1, but got 0 argument(s) instead."
    );

    expect(sanitizer([1])).toEqual([1]);
    expect(sanitizer([1, false])).toEqual([1, 0]);
    expect(sanitizer(["-1", 2, true])).toEqual([-1, 2, 1]);
    expect(sanitizer(["-1", 2, undefined, true])).toEqual([-1, 2, 0, 1]);
    expect(sanitizer(["-1", 2, true, undefined])).toEqual([-1, 2, 1, 0]);
  });

  test("repeating, optional, number argument", () => {
    const argList = args`n (number,repeating,optional) some number`;

    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    expect(sanitizer([])).toEqual([]);
    expect(sanitizer([1])).toEqual([1]);
    expect(sanitizer([1, false])).toEqual([1, 0]);
    expect(sanitizer(["-1", 2, true])).toEqual([-1, 2, 1]);
    expect(sanitizer(["-1", 2, undefined, true])).toEqual([-1, 2, 0, 1]);
    expect(sanitizer(["-1", 2, true, undefined])).toEqual([-1, 2, 1, 0]);
  });

  test("an optional number argument after another argument", () => {
    const argList = args`
      m (number) a number
      n (number,optional) another number`;

    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    expect(sanitizer([1])).toEqual([1]);
    expect(sanitizer([1, false])).toEqual([1, 0]);
    expect(sanitizer([false, true])).toEqual([0, 1]);
    expect(sanitizer([1, undefined])).toEqual([1, undefined]);
  });

  test("a single boolean argument", () => {
    const argList = args`b (boolean) some boolean value`;

    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    expect(sanitizer([1])).toEqual([true]);
    expect(sanitizer([0])).toEqual([false]);
    expect(sanitizer([true])).toEqual([true]);
    expect(sanitizer([false])).toEqual([false]);
    expect(sanitizer([undefined])).toEqual([false]);
    expect(sanitizer([""])).toEqual([false]);
    expect(sanitizer(["false"])).toEqual([false]);
    expect(sanitizer(["true"])).toEqual([true]);
    expect(sanitizer(["TRUE"])).toEqual([true]);
    expect(() => sanitizer(["abc"])).toThrow(
      'Argument "b" should be a boolean, but "abc" is a text, and cannot be coerced to a boolean.'
    );
    expect(() => sanitizer(["1"])).toThrow(
      'Argument "b" should be a boolean, but "1" is a text, and cannot be coerced to a boolean.'
    );
  });

  test("a single string argument", () => {
    const argList = args`s (string) some string`;

    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    expect(sanitizer(["abc"])).toEqual(["abc"]);
    expect(sanitizer([1])).toEqual(["1"]);
    expect(sanitizer([false])).toEqual(["FALSE"]);
    expect(sanitizer([true])).toEqual(["TRUE"]);
    expect(sanitizer([undefined])).toEqual([""]);
  });

  test("a simple untyped range argument", () => {
    const argList = args`r (range) some values`;
    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    const m1: Range<number> = [[1, 2]]; // one column with 1 2
    expect(sanitizer([m1])).toEqual([m1]);

    const m2 = [
      [undefined, undefined],
      ["abc", true]
    ];
    expect(sanitizer([m2])).toEqual([m2]);

    expect(() => sanitizer([1])).toThrow('Argument "r" has the wrong type');
  });

  test("a simple range of numbers", () => {
    const argList = args`r (range<number>) some numeric values`;
    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    const m1 = [[1, 2]]; // one column with 1 2
    expect(sanitizer([m1])).toEqual([m1]);

    const m2 = [
      [undefined, 3],
      ["abc", 1]
    ];
    const m2_number = [
      [undefined, 3],
      [undefined, 1]
    ];
    expect(sanitizer([m2])).toEqual([m2_number]);

    expect(() => sanitizer([1])).toThrow('Argument "r" has the wrong type');
  });

  test("a simple range of booleans", () => {
    const argList = args`r (range<boolean>) some boolean values`;
    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    const m1: Range<boolean> = [[true, false]];
    expect(sanitizer([m1])).toEqual([m1]);

    const m2 = [
      [undefined, 3],
      [true, "true"]
    ];
    const m2_number = [
      [undefined, undefined],
      [true, undefined]
    ];
    expect(sanitizer([m2])).toEqual([m2_number]);

    expect(() => sanitizer([1])).toThrow('Argument "r" has the wrong type');
  });

  test("a simple range of strings", () => {
    const argList = args`r (range<string>) some string values`;
    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    const m1 = [["a", "b"]]; // one column with 1 2
    expect(sanitizer([m1])).toEqual([m1]);

    const m2 = [
      [undefined, 3],
      ["abc", 1]
    ];
    const m2_number = [
      [undefined, undefined],
      ["abc", undefined]
    ];
    expect(sanitizer([m2])).toEqual([m2_number]);

    expect(() => sanitizer([1])).toThrow('Argument "r" has the wrong type');
  });

  test("number or range of numbers", () => {
    const argList = args`n (number,range<number>) some numeric values`;
    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    expect(sanitizer([1])).toEqual([1]);
    expect(sanitizer([false])).toEqual([0]);
    expect(sanitizer([true])).toEqual([1]);
    expect(sanitizer([undefined])).toEqual([0]);
    expect(sanitizer([""])).toEqual([0]);
    expect(sanitizer(["1"])).toEqual([1]);
    expect(sanitizer(["-1"])).toEqual([-1]);
    expect(sanitizer(["1.1"])).toEqual([1.1]);

    expect(() => sanitizer(["ab"])).toThrow(
      'Argument "n" should be a number, but "ab" is a text, and cannot be coerced to a number.'
    );
    expect(() => sanitizer([])).toThrow(
      "Wrong number of arguments. Expected 1, but got 0 argument(s) instead."
    );
    expect(() => sanitizer([1, 2])).toThrow(
      "Wrong number of arguments. Expected 1, but got 2 argument(s) instead."
    );

    const m1 = [[1, 2]]; // one column with 1 2
    expect(sanitizer([m1])).toEqual([m1]);

    const m2 = [
      [undefined, 3],
      ["abc", 1]
    ];
    const m2_number = [
      [undefined, 3],
      [undefined, 1]
    ];
    expect(sanitizer([m2])).toEqual([m2_number]);
  });

  test("string or range of strings", () => {
    const argList = args`s (string,range<string>) some string values`;
    const sanitizer = makeSanitizer(argList);
    expect(sanitizer.toString()).toMatchSnapshot();

    expect(sanitizer(["abc"])).toEqual(["abc"]);
    expect(sanitizer([1])).toEqual(["1"]);
    expect(sanitizer([false])).toEqual(["FALSE"]);
    expect(sanitizer([true])).toEqual(["TRUE"]);
    expect(sanitizer([undefined])).toEqual([""]);

    const m1 = [["a", "b"]]; // one column with 1 2
    expect(sanitizer([m1])).toEqual([m1]);

    const m2 = [
      [undefined, 3],
      ["abc", 1]
    ];
    const m2_number = [
      [undefined, undefined],
      ["abc", undefined]
    ];
    expect(sanitizer([m2])).toEqual([m2_number]);
  });
});
