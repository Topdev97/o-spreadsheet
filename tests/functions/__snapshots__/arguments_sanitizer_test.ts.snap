// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeSanitizer a simple range of booleans 1`] = `
"function anonymous(arg0
) {
if (arguments.length !== 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
if (!(arg0 instanceof Array)) {
   throw new Error(\`Argument \\"r\\" has the wrong type\`);
}
for (let i = 0; i < arg0.length; i++) {
  let col = arg0[i];
  for (let j = 0; j < col.length; j++) {
    if (typeof col[j] !== \\"boolean\\") {
      col[j] = undefined;
    }
  }
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer a simple range of numbers 1`] = `
"function anonymous(arg0
) {
if (arguments.length !== 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
if (!(arg0 instanceof Array)) {
   throw new Error(\`Argument \\"r\\" has the wrong type\`);
}
for (let i = 0; i < arg0.length; i++) {
  let col = arg0[i];
  for (let j = 0; j < col.length; j++) {
    if (typeof col[j] !== \\"number\\") {
      col[j] = undefined;
    }
  }
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer a simple range of strings 1`] = `
"function anonymous(arg0
) {
if (arguments.length !== 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
if (!(arg0 instanceof Array)) {
   throw new Error(\`Argument \\"r\\" has the wrong type\`);
}
for (let i = 0; i < arg0.length; i++) {
  let col = arg0[i];
  for (let j = 0; j < col.length; j++) {
    if (typeof col[j] !== \\"string\\") {
      col[j] = undefined;
    }
  }
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer a simple untyped range argument 1`] = `
"function anonymous(arg0
) {
if (arguments.length !== 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
if (!(arg0 instanceof Array)) {
   throw new Error(\`Argument \\"r\\" has the wrong type\`);
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer a single boolean argument 1`] = `
"function anonymous(arg0
) {
if (arguments.length !== 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
switch (typeof arg0) {
 case \\"undefined\\":
   arg0 = false;
   break;
 case \\"number\\":
   arg0 = arg0 ? true : false;
   break;
 case \\"string\\":
   if (arg0) {
     let uppercaseVal = arg0.toUpperCase();
     if (uppercaseVal === \\"TRUE\\") {
       arg0 = true;
     } else if (uppercaseVal === \\"FALSE\\") {
       arg0 = false;
     } else {
       throw new Error(\`Argument \\"b\\" should be a boolean, but \\"\${arg0}\\" is a text, and cannot be coerced to a boolean.\`);
     }
   } else {
     arg0 = false;
   }
   break;
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer a single number argument 1`] = `
"function anonymous(arg0
) {
if (arguments.length !== 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
switch (typeof arg0) {
 case \\"undefined\\":
   arg0 = 0;
   break;
 case \\"boolean\\":
   arg0 = arg0 ? 1 : 0;
   break;
 case \\"string\\":
   if (arg0) {
     let n = Number(arg0);
     if (isNaN(n)) {
       throw new Error(\`Argument \\"n\\" should be a number, but \\"\${arg0}\\" is a text, and cannot be coerced to a number.\`);
     } else {
       arg0 = n;
     }
   } else {
     arg0 = 0;
   }
   break;
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer a single string argument 1`] = `
"function anonymous(arg0
) {
if (arguments.length !== 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
switch (typeof arg0) {
 case \\"undefined\\":
   arg0 = \\"\\";
   break;
 case \\"number\\":
   arg0 = arg0.toString();
   break;
 case \\"boolean\\":
   arg0 = arg0 ? \\"TRUE\\" : \\"FALSE\\";
   break;
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer an optional number argument 1`] = `
"function anonymous(arg0
) {
if (arguments.length > 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
switch (typeof arg0) {
 case \\"boolean\\":
   arg0 = arg0 ? 1 : 0;
   break;
 case \\"string\\":
   if (arg0) {
     let n = Number(arg0);
     if (isNaN(n)) {
       throw new Error(\`Argument \\"n\\" should be a number, but \\"\${arg0}\\" is a text, and cannot be coerced to a number.\`);
     } else {
       arg0 = n;
     }
   } else {
     arg0 = 0;
   }
   break;
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer an optional number argument after another argument 1`] = `
"function anonymous(arg0,arg1
) {
if (arguments.length < 1 || arguments.length > 2) {
  throw new Error(\`Wrong number of arguments. Expected 2, but got \${arguments.length} argument(s) instead.\`);
}
switch (typeof arg0) {
 case \\"undefined\\":
   arg0 = 0;
   break;
 case \\"boolean\\":
   arg0 = arg0 ? 1 : 0;
   break;
 case \\"string\\":
   if (arg0) {
     let n = Number(arg0);
     if (isNaN(n)) {
       throw new Error(\`Argument \\"m\\" should be a number, but \\"\${arg0}\\" is a text, and cannot be coerced to a number.\`);
     } else {
       arg0 = n;
     }
   } else {
     arg0 = 0;
   }
   break;
}
switch (typeof arg1) {
 case \\"boolean\\":
   arg1 = arg1 ? 1 : 0;
   break;
 case \\"string\\":
   if (arg1) {
     let n = Number(arg1);
     if (isNaN(n)) {
       throw new Error(\`Argument \\"n\\" should be a number, but \\"\${arg1}\\" is a text, and cannot be coerced to a number.\`);
     } else {
       arg1 = n;
     }
   } else {
     arg1 = 0;
   }
   break;
}
return this.fn(arg0,arg1);
}"
`;

exports[`makeSanitizer an optional number argument with a default value 1`] = `
"function anonymous(arg0
) {
if (arguments.length > 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
switch (typeof arg0) {
 case \\"undefined\\":
   arg0 = 42;
   break;
 case \\"boolean\\":
   arg0 = arg0 ? 1 : 0;
   break;
 case \\"string\\":
   if (arg0) {
     let n = Number(arg0);
     if (isNaN(n)) {
       throw new Error(\`Argument \\"n\\" should be a number, but \\"\${arg0}\\" is a text, and cannot be coerced to a number.\`);
     } else {
       arg0 = n;
     }
   } else {
     arg0 = 0;
   }
   break;
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer number or range of numbers 1`] = `
"function anonymous(arg0
) {
if (arguments.length !== 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
switch (typeof arg0) {
 case \\"undefined\\":
   arg0 = 0;
   break;
 case \\"boolean\\":
   arg0 = arg0 ? 1 : 0;
   break;
 case \\"string\\":
   if (arg0) {
     let n = Number(arg0);
     if (isNaN(n)) {
       throw new Error(\`Argument \\"n\\" should be a number, but \\"\${arg0}\\" is a text, and cannot be coerced to a number.\`);
     } else {
       arg0 = n;
     }
   } else {
     arg0 = 0;
   }
   break;
  case \\"object\\":
if (!(arg0 instanceof Array)) {
   throw new Error(\`Argument \\"n\\" has the wrong type\`);
}
for (let i = 0; i < arg0.length; i++) {
  let col = arg0[i];
  for (let j = 0; j < col.length; j++) {
    if (typeof col[j] !== \\"number\\") {
      col[j] = undefined;
    }
  }
}
    break;
}
return this.fn(arg0);
}"
`;

exports[`makeSanitizer repeating, non optional, number argument 1`] = `
"function anonymous(...arg0
) {
if (arguments.length < 1) {
  throw new Error(\`Wrong number of arguments. Expected 1, but got \${arguments.length} argument(s) instead.\`);
}
for (let i = 0; i < arg0.length; i++) {
let arg0_value = arg0[i];
switch (typeof arg0_value) {
 case \\"undefined\\":
   arg0_value = 0;
   break;
 case \\"boolean\\":
   arg0_value = arg0_value ? 1 : 0;
   break;
 case \\"string\\":
   if (arg0_value) {
     let n = Number(arg0_value);
     if (isNaN(n)) {
       throw new Error(\`Argument \\"n\\" should be a number, but \\"\${arg0_value}\\" is a text, and cannot be coerced to a number.\`);
     } else {
       arg0_value = n;
     }
   } else {
     arg0_value = 0;
   }
   break;
}
arg0[i] = arg0_value;
}
return this.fn(...arg0);
}"
`;

exports[`makeSanitizer repeating, optional, number argument 1`] = `
"function anonymous(...arg0
) {
for (let i = 0; i < arg0.length; i++) {
let arg0_value = arg0[i];
switch (typeof arg0_value) {
 case \\"undefined\\":
   arg0_value = 0;
   break;
 case \\"boolean\\":
   arg0_value = arg0_value ? 1 : 0;
   break;
 case \\"string\\":
   if (arg0_value) {
     let n = Number(arg0_value);
     if (isNaN(n)) {
       throw new Error(\`Argument \\"n\\" should be a number, but \\"\${arg0_value}\\" is a text, and cannot be coerced to a number.\`);
     } else {
       arg0_value = n;
     }
   } else {
     arg0_value = 0;
   }
   break;
}
arg0[i] = arg0_value;
}
return this.fn(...arg0);
}"
`;
